crates/globwalk/src/lib.rs

```diff
@@ -0,0 +1,129 @@
+use std::fmt::Display;
+
+use glob_match::glob_match;
+use turborepo_paths::AbsolutePathBuf;
+
+pub enum WalkType {
+    Files,
+    Folders,
+    All,
+}
+
+#[derive(Debug)]
+pub enum WalkError {
+    Error(walkdir::Error),
+}
+
+impl Display for WalkError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            WalkError::Error(e) => e.fmt(f),
+        }
+    }
+}
+
+impl std::error::Error for WalkError {}
+
+/// Performs a glob walk, yielding paths that
+/// _are_ included in the include list (if it is nonempty)
+/// and _not_ included in the exclude list.
+///
+/// In the case of an empty include, then all
+/// files are included.
+pub fn globwalk<'a>(
+    base_path: AbsolutePathBuf,
+    include: &'a [String],
+    exclude: &'a [String],
+    walk_type: WalkType,
+) -> impl Iterator<Item = Result<AbsolutePathBuf, WalkError>> + 'a {
+    let walker = walkdir::WalkDir::new(base_path.as_path());
+
+    walker
+        .into_iter()
+        // we need to eagerly filter folders, preventing traversal into
+        // folders that aren't included or are excluded. we do this by
+        // evaluating 'potential matches' for the include list meaning
+        // 'it is possible that a subfolder may match', as well as definite
+        // matches for the exclude list
+        .filter_entry(|e| {
+            include
+                .iter()
+                .any(|p| potential_match(e.path().to_string_lossy().as_ref(), p))
```

Comment:
Maybe we should ignore or error on non-Unicode paths

Reply:
Going to bubble up these issues instead of lossy consumption.